docs: add product-landing.json authoring details (#190)
hashicorp/dev-portal: src/data/README.md

  "product": "waypoint",
  "heading": "Introduction to Waypoint",
  "text": "Welcome to Waypoint! This introduction section covers what Waypoint is, the problem Waypoint aims to solve, and how Waypoint compares to other software.",
  "link": {
    "text": "Get Started",
    "url": "/waypoint/docs/intro"
  }
}
```

![](https://user-images.githubusercontent.com/4624598/158821262-03798dca-12e6-487b-ac3e-e8bab51be8b1.png)

</details>

<!-- block type: cards -->

<details>
<summary>Block type: <code>"cards"</code></summary>

The `"cards"` block displays a grid of `CardLink`s, each linked using a single `url`.

| Property | Type | Details |
| --- | --- | --- |
| `type` | `"cards"` | Block type |
| `columns` | `2 \| 3` | The maximum number of columns |
| `cards` | `Array<{ icon, iconBrandColor, heading, text, url, tags }>` | An array of objects, described in detail below |

Each item in the `cards` array has the following structure:

| Property | Type | Details |
| --- | --- | --- |
| `icon` | (optional) `string` | Optional icon to show at the top of the card. Must be one of the keys in [the card component's icon dictionary](../views/product-landing/components/cards/icon-dict.js) |

> | `iconBrandColor` | (optional) [Products](https://github.com/hashicorp/web-platform-packages/blob/69b2cf609a254c0f469142266ecb60ffe04cbc7a/packages/product-meta/index.tsx#L11) string or `"neutral" \| "neutral-dark"` | Optional brand color override to apply to the icon. Defaults to the current product context. |

Reviewer: We should link to `ProductSlug` type here since it's built on top of what you linked, and adds `sentinel` and `hcp` slugs.

```suggestion
| `iconBrandColor` | (optional) [ProductSlug](/src/types/products.ts) string or `"neutral" \| "neutral-dark"` | Optional brand color override to apply to the icon. Defaults to the current product context. |
```

Author: I think `Products` is the correct type here - passing `"hcp"` or `"sentinel"` might not work as expected, since they're not supported options in `IconTile`. Maybe more clear if I link to [`ProductBrandColor` within `IconTile`'s `types.ts` file](https://github.com/hashicorp/dev-portal/blob/476bd48cbd0b28bcbe86d45af758cbbad66ee807/src/components/icon-tile/types.ts#L3), which is equivalent but declared within the project and more direct (since this is ultimately getting passed to `IconTile`). I just made that update, let me know what you think!

Reviewer: Ah okay! I misremembered whether _IconTile_ handled those two or not. But I like your idea of linking to `ProductBrandColor` even more! 👏🏻

This file is only used in DevDot. It contains the data and content needed for each product's landing page (`/boundary`, `/consul`, etc.).

### Editing product landing content

Our `<product slug>-landing.json` pages take a "block"-based authoring approach for the main content area. This approach allows authors to write content in JSON instead of JSX. As an overview, each file has three properties, all are required:

```json5
{

>   "heading": "string", // required
>   "subheading": "string", // required
>   "blocks": [ /*  array of objects, required */ ], 

Reviewer: Since you note right before this that the properties are required, I think you can leave off the comments about it here. Also found a sneaky extra space character! 😊

```suggestion
  "heading": "string",
  "subheading": "string",
  "blocks": [ /* array of objects */ ], 
```

}
```

Each of these top-level properties is addressed in more detail below.

#### `heading` and `subheading`

Set the heading and subheading shown on the page.

<details>
<summary>Example</summary>

Source: 

> 

Reviewer: Extra line here

```suggestion
```

```json5
{
  "heading": "Waypoint Documentation", 
  "subheading": "Use Waypoint to deliver a PaaS-like experience for Kubernetes, ECS, and other platforms.", 
  "blocks": [ /* ... */ ],
}
```

Result:

![](https://user-images.githubusercontent.com/4624598/158818382-e78ea677-85c1-41aa-92b4-ca8714f06f2d.png)

</details>

#### `blocks`

Each item in the `blocks` array represents a component on the page. Each of these items must have a `type` property, which can be one of the types listed below.

> <!-- block: type heading -->

Reviewer: These comments are great when reading the raw Markdown! Nice touch ✨

<details>
<summary>Block type: <code>"heading"</code></summary>

Heading blocks render HTML heading elements. Each block accepts the following properties:

| Property | Type | Details |
| --- | --- | --- |
| `type` | `"heading"` | Block type |
| `heading` | `string` | Text for the heading |

> | `level` | `2 \| 3 \| 4 \| 5 \| 6` | Semantic heading level, for example `2` becomes `<h2>`. Note that there is already an `<h1>` rendered for the page, so only values  should be used. |

Reviewer: Missing word here, I think `these`?

```suggestion
| `level` | `2 \| 3 \| 4 \| 5 \| 6` | Semantic heading level, for example `2` becomes `<h2>`. Note that there is already an `<h1>` rendered for the page, so only these values should be used. |
```

> | `size` | `100 \| 200 \| 300 \| 400 \| 500` | Visual size of the heading. `500` is the largest size and `100` is the smallest. Visual size should generally reflect the semantic level, with `h2 = 300`, `h3 = 200`, and `h4` and below at the `100` size.

Reviewer: 💭 Might be worth linking to the `Heading` component here?

Author: ✅ Linked!

Example: `h2` with `300` size:

```json5
{
  "type": "heading",
  "heading": "Featured Reference Docs",
  "level": 2,
  "size": 300,
}
```

![](https://user-images.githubusercontent.com/4624598/158818745-a20d1892-efa6-4053-9d00-811645d642aa.png)

Example: `h2` with `400` size:

```json5
{
  "type": "heading",
  "heading": "Explore Waypoint Documentation",
  "level": 2,
  "size": 400,
}
```

![](https://user-images.githubusercontent.com/4624598/158818617-2b8ce029-ad41-4081-8701-869d51abf40b.png)

</details>

<!-- block type: get_started -->

<details>
<summary>Block type: <code>"get_started"</code></summary>

The `"get_started"` block renders a heading, descriptive text, and a single link alongside a product icon.

| Property | Type | Details |
| --- | --- | --- |
| `type` | `"get_started"` | Block type |
| `product` | `"boundary" \| "consul" \| "nomad" \| "packer" \| "terraform" \| "vault" \| "vagrant" \| "waypoint" \| "sentinel" \| "hcp"` | Product icon to be shown. |
| `heading` | `string` | Text for the heading |
| `text` | `string` | Descriptive text shown below the heading |

> | `link` | `{ text: string, url: string }` | Link shown below the body text |

Reviewer: 💭 Do you think referencing the `StandaloneLink` here is too much info?

```suggestion
| `link` | `{ text: string, url: string }` | `StandaloneLink` shown below the body text |
```

Author: I think it's a good suggestion! Still clear to authors that it's a link, clearer to us that it's a specific component being used 👍

Example: Waypoint `"get_started"` block

```json5
{
  "type": "get_started",
  "product": "waypoint",
  "heading": "Introduction to Waypoint",
  "text": "Welcome to Waypoint! This introduction section covers what Waypoint is, the problem Waypoint aims to solve, and how Waypoint compares to other software.",
  "link": {
    "text": "Get Started",
    "url": "/waypoint/docs/intro"
  }
}
```

![](https://user-images.githubusercontent.com/4624598/158821262-03798dca-12e6-487b-ac3e-e8bab51be8b1.png)

</details>

<!-- block type: cards -->

<details>
<summary>Block type: <code>"cards"</code></summary>

> The `"cards"` block displays a grid of cards, each linked using a single `url`.

Reviewer: 💭 Similar thoughts here as previous comment, do you think it would be valuable to specify `CardLink` here?

```suggestion
The `"cards"` block displays a grid of `CardLink`s, each linked using a single `url`.
```

Author: I dig it 👍

| Property | Type | Details |
| --- | --- | --- |
| `type` | `"cards"` | Block type |
| `columns` | `2 \| 3` | The maximum number of columns |

> | `cards` | `Array<{ icon, iconBrandColor, heading, text, url, tags }>` | An array of `card` objects, described in detail below |

Reviewer: 💭 Not sure if `card` is needed here since it's not the name of a property or TypeScript object at the moment.

```suggestion
| `cards` | `Array<{ icon, iconBrandColor, heading, text, url, tags }>` | An array of objects, described in detail below |
```

Each item in the `cards` array has the following structure:

| Property | Type | Details |
| --- | --- | --- |
| `icon` | (optional) `"IconBox" \| "IconConsulColor" \| "IconDocs" \| "IconDownload" \| "IconPackerColor" \| "IconProvider" \| "IconServer" \| "IconTerminal" \| "IconTerraformColor" \| "IconVaultColor"` | Optional icon to show at the top of the card. |
| `iconBrandColor` | (optional) `"boundary" \| "consul" \| "nomad" \| "packer" \| "terraform" \| "vault" \| "vagrant" \| "waypoint" \| "neutral" \| "neutral-dark"` | Optional brand color override to apply to the icon. Defaults to the current product context. |
| `heading` | `string` | Text for the card heading |
| `text` | `string` | Text for the card body |
| `url` | `string` | URL to link to |

> | `tags` | (optional) `Array<"boundary" \| "consul" \| "nomad" \| "packer" \| "terraform" \| "vagrant" \| "vault" \| "video" \| "waypoint">` | Optional array of tags, to be displayed as icons at the bottom of the card |

Reviewer: 💭 What do you think about linking to the TypeScript where `icon`, `iconBrandColor`, and `tags` are defined instead of listing them out? I worry we'll forget to update these docs if we add or remove options for them. 😅

Author: Ooh yeah I like that - will update! 👍

Author: ✅  updated!
